{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d157d830",
   "metadata": {},
   "source": [
    "# Getting Chr5 novel node statistics\n",
    "\n",
    "This is more or less following the hpgp-basics doc, but through python rather than R.\n",
    "\n",
    "First, we need to import stuff:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35bcc145",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['4871018.txt', '3443998.txt', '1330802.txt']\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "\n",
    "node_dir = \"nonref_files/chr5-1k-10-50/\"\n",
    "nodes = os.listdir(node_dir)\n",
    "\n",
    "sample_data = pd.read_csv(\"nonref_files/20130606_g1k_3202_samples_ped_population.txt\", \n",
    "                         delimiter=\" \")\n",
    "sample_data[\"SampleID\"] = sample_data[\"SampleID\"].astype(str)\n",
    "\n",
    "print(nodes[:3])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cfd8699",
   "metadata": {},
   "source": [
    "Now we need to get all the path information for each node, and convert it into a dataframe so we can do something useful with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "883de34e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "pop_counts = {}\n",
    "node_data = \"\"\n",
    "nid = \"\"\n",
    "\n",
    "for node in nodes[:]: # only working on the first node for now\n",
    "    if not node.endswith(\".txt\") : continue\n",
    "    nid = node[:node.find(\".\")]\n",
    "    paths = open(node_dir + node).read().split('\\n')[:-2] # 2 trailing newlines\n",
    "    \n",
    "    counts = []\n",
    "    # converting to a set makes sure homozygotes are only counted once\n",
    "    for path in sorted(list(set(paths))):\n",
    "        counts.append((path, paths.count(path)))\n",
    "    \n",
    "    # grab population counts\n",
    "    test = pd.DataFrame(counts, columns=[\"SampleID\", \"count\"])\n",
    "    test[\"SampleID\"] = test[\"SampleID\"].astype(str)\n",
    "    test = test.merge(sample_data[[\"SampleID\", \"Population\"]], on=\"SampleID\")\n",
    "    pop_counts[nid] = test[\"Population\"].value_counts()\n",
    "    \n",
    "    # we can work on the whole thing later, but let's just grab one node for now.\n",
    "    node_data = test\n",
    "     \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "550705e6",
   "metadata": {},
   "source": [
    "We can now use `node_data` to get information about the populations in our node."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "184cc8c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counts for 4971695:\n",
      "Population\n",
      "GWD    5\n",
      "PEL    3\n",
      "PUR    2\n",
      "MSL    2\n",
      "CHS    1\n",
      "ESN    1\n",
      "ASW    1\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(f\"Counts for {nid}:\")\n",
    "print(node_data[\"Population\"].value_counts())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63855f5c",
   "metadata": {},
   "source": [
    "We can see that most of the paths represented are part of the GWD, PUR, and MSL populations.\n",
    "\n",
    "We can then grab the path prefixes "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1d986ade",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13    Gambian in Western Division, The Gambia\n",
      "Name: Population Description, dtype: object\n",
      "   SampleID  count Population\n",
      "6   HG02572      1        GWD\n",
      "7   HG02622      1        GWD\n",
      "8   HG02630      1        GWD\n",
      "9   HG02717      1        GWD\n",
      "10  HG02886      2        GWD\n"
     ]
    }
   ],
   "source": [
    "populations = pd.read_csv(\"data/20131219.populations.tsv\", delimiter=\"\\t\")\n",
    "print(populations[populations[\"Population Code\"] == \"GWD\"][\"Population Description\"])\n",
    "\n",
    "\n",
    "\n",
    "print(node_data[node_data[\"Population\"]==\"GWD\"])\n",
    "\n",
    "path_prefixes = node_data[node_data[\"Population\"]==\"GWD\"][\"SampleID\"].tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36be5ad9",
   "metadata": {},
   "source": [
    "If we want to, we can then extract just the node (and the ones surrounding for convenience/if we want to use bandage/etc to visualise later). In theory we could then use this to visually determine if there was any variation within this node (?)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5fd79cbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "node_og = f\"nonref_files/chr5-1k-10-50/node_{nid}.og\"\n",
    "\n",
    "subprocess.run([\"./bin/odgi\", \"extract\", \"-i\", \"chr5.full.og\", \"-n\", nid,\n",
    "                \"-o\", node_og, \"-t\", \"4\"]);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "59affae8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HG02572#1#JAHAOW010000129.1#0:72387-73390', 'HG02622#2#JAHAON010000014.1#0:104211456-104212459', 'HG02630#1#JAHAOQ010000041.1#0:14798010-14799013', 'HG02717#1#JAHAOS010000011.1#0:27717791-27718794', 'HG02886#1#JAHAOU010000019.1#0:473261-474264', 'HG02886#2#JAHAOT010000128.1#0:11943547-11944550']\n"
     ]
    }
   ],
   "source": [
    "paths = subprocess.run([\"./bin/odgi\", \"paths\", \"-i\", node_og,\n",
    "                       \"-L\"], capture_output=True).stdout.decode().split()\n",
    "        \n",
    "paths = [p for p in paths if p[:p.find(\"#\")] in path_prefixes]\n",
    "print(paths)\n",
    "\n",
    "f = open(\"_paths.txt\", \"w\") #remove later\n",
    "\n",
    "for p in paths:\n",
    "    f.write(p + \"\\n\")\n",
    "\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb81a79b",
   "metadata": {},
   "source": [
    "## Looking at Populations\n",
    "\n",
    "Let's have a look at populations - by doing some data transformation, we can see how many paths of each population are in each node.\n",
    "\n",
    "Let's also get an overview of the population codes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1ff01fb8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Population Description</th>\n",
       "      <th>Population Code</th>\n",
       "      <th>Super Population</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Chinese Dai in Xishuangbanna, China</td>\n",
       "      <td>CDX</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Han Chinese in Bejing, China</td>\n",
       "      <td>CHB</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Japanese in Tokyo, Japan</td>\n",
       "      <td>JPT</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Kinh in Ho Chi Minh City, Vietnam</td>\n",
       "      <td>KHV</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Southern Han Chinese, China</td>\n",
       "      <td>CHS</td>\n",
       "      <td>EAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Bengali in Bangladesh</td>\n",
       "      <td>BEB</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Gujarati Indian in Houston,TX</td>\n",
       "      <td>GIH</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Indian Telugu in the UK</td>\n",
       "      <td>ITU</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Punjabi in Lahore,Pakistan</td>\n",
       "      <td>PJL</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Sri Lankan Tamil in the UK</td>\n",
       "      <td>STU</td>\n",
       "      <td>SAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>African Ancestry in Southwest US</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>African Caribbean in Barbados</td>\n",
       "      <td>ACB</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Esan in Nigeria</td>\n",
       "      <td>ESN</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>Gambian in Western Division, The Gambia</td>\n",
       "      <td>GWD</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>Luhya in Webuye, Kenya</td>\n",
       "      <td>LWK</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>Mende in Sierra Leone</td>\n",
       "      <td>MSL</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>Yoruba in Ibadan, Nigeria</td>\n",
       "      <td>YRI</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>British in England and Scotland</td>\n",
       "      <td>GBR</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>Finnish in Finland</td>\n",
       "      <td>FIN</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>Iberian populations in Spain</td>\n",
       "      <td>IBS</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>Toscani in Italy</td>\n",
       "      <td>TSI</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>Utah residents with Northern and Western Europ...</td>\n",
       "      <td>CEU</td>\n",
       "      <td>EUR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>Colombian in Medellin, Colombia</td>\n",
       "      <td>CLM</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>Mexican Ancestry in Los Angeles, California</td>\n",
       "      <td>MXL</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>Peruvian in Lima, Peru</td>\n",
       "      <td>PEL</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>Puerto Rican in Puerto Rico</td>\n",
       "      <td>PUR</td>\n",
       "      <td>AMR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               Population Description Population Code  \\\n",
       "0                 Chinese Dai in Xishuangbanna, China             CDX   \n",
       "1                        Han Chinese in Bejing, China             CHB   \n",
       "2                            Japanese in Tokyo, Japan             JPT   \n",
       "3                   Kinh in Ho Chi Minh City, Vietnam             KHV   \n",
       "4                         Southern Han Chinese, China             CHS   \n",
       "5                               Bengali in Bangladesh             BEB   \n",
       "6                       Gujarati Indian in Houston,TX             GIH   \n",
       "7                             Indian Telugu in the UK             ITU   \n",
       "8                          Punjabi in Lahore,Pakistan             PJL   \n",
       "9                          Sri Lankan Tamil in the UK             STU   \n",
       "10                   African Ancestry in Southwest US             ASW   \n",
       "11                      African Caribbean in Barbados             ACB   \n",
       "12                                    Esan in Nigeria             ESN   \n",
       "13            Gambian in Western Division, The Gambia             GWD   \n",
       "14                             Luhya in Webuye, Kenya             LWK   \n",
       "15                              Mende in Sierra Leone             MSL   \n",
       "16                          Yoruba in Ibadan, Nigeria             YRI   \n",
       "17                    British in England and Scotland             GBR   \n",
       "18                                 Finnish in Finland             FIN   \n",
       "19                       Iberian populations in Spain             IBS   \n",
       "20                                   Toscani in Italy             TSI   \n",
       "21  Utah residents with Northern and Western Europ...             CEU   \n",
       "22                    Colombian in Medellin, Colombia             CLM   \n",
       "23        Mexican Ancestry in Los Angeles, California             MXL   \n",
       "24                             Peruvian in Lima, Peru             PEL   \n",
       "25                        Puerto Rican in Puerto Rico             PUR   \n",
       "\n",
       "   Super Population  \n",
       "0               EAS  \n",
       "1               EAS  \n",
       "2               EAS  \n",
       "3               EAS  \n",
       "4               EAS  \n",
       "5               SAS  \n",
       "6               SAS  \n",
       "7               SAS  \n",
       "8               SAS  \n",
       "9               SAS  \n",
       "10              AFR  \n",
       "11              AFR  \n",
       "12              AFR  \n",
       "13              AFR  \n",
       "14              AFR  \n",
       "15              AFR  \n",
       "16              AFR  \n",
       "17              EUR  \n",
       "18              EUR  \n",
       "19              EUR  \n",
       "20              EUR  \n",
       "21              EUR  \n",
       "22              AMR  \n",
       "23              AMR  \n",
       "24              AMR  \n",
       "25              AMR  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "populations[[\"Population Description\", \"Population Code\", \"Super Population\"]].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f7dc561d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Population</th>\n",
       "      <th>node_id</th>\n",
       "      <th>ACB</th>\n",
       "      <th>ASW</th>\n",
       "      <th>CHS</th>\n",
       "      <th>CLM</th>\n",
       "      <th>ESN</th>\n",
       "      <th>GWD</th>\n",
       "      <th>KHV</th>\n",
       "      <th>MSL</th>\n",
       "      <th>PEL</th>\n",
       "      <th>PJL</th>\n",
       "      <th>PUR</th>\n",
       "      <th>YRI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1030533</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1030773</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1031072</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1330802</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1330842</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Population  node_id  ACB  ASW  CHS  CLM  ESN  GWD  KHV  MSL  PEL  PJL  PUR  \\\n",
       "0           1030533  2.0  1.0  1.0  1.0  0.0  2.0  0.0  2.0  1.0  0.0  2.0   \n",
       "1           1030773  2.0  1.0  1.0  1.0  0.0  2.0  0.0  2.0  1.0  0.0  2.0   \n",
       "2           1031072  2.0  1.0  1.0  1.0  0.0  2.0  0.0  2.0  1.0  0.0  3.0   \n",
       "3           1330802  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  2.0  0.0  2.0   \n",
       "4           1330842  0.0  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0  2.0   \n",
       "\n",
       "Population  YRI  \n",
       "0           1.0  \n",
       "1           1.0  \n",
       "2           1.0  \n",
       "3           0.0  \n",
       "4           0.0  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# some completely unnecessary data transformation - may be interesting later...    \n",
    "node_pop_df = pd.DataFrame.from_dict(pop_counts).reset_index()\n",
    "node_pop_df = pd.pivot_table(node_pop_df, columns=\"Population\").reset_index(names=\"node_id\").fillna(0)\n",
    "\n",
    "node_pop_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4748ea02",
   "metadata": {},
   "source": [
    "The natural next step for this is to perform the same operation across every chromosome :)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e7756a8",
   "metadata": {},
   "source": [
    "## (In theory, Leading to the admixture stuff)\n",
    "\n",
    "We can also sort by paths of a given population:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "68d53fe0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Population</th>\n",
       "      <th>node_id</th>\n",
       "      <th>ACB</th>\n",
       "      <th>ASW</th>\n",
       "      <th>CHS</th>\n",
       "      <th>CLM</th>\n",
       "      <th>ESN</th>\n",
       "      <th>GWD</th>\n",
       "      <th>KHV</th>\n",
       "      <th>MSL</th>\n",
       "      <th>PEL</th>\n",
       "      <th>PJL</th>\n",
       "      <th>PUR</th>\n",
       "      <th>YRI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1030533</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1030773</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>4960359</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>4960603</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>4971420</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Population  node_id  ACB  ASW  CHS  CLM  ESN  GWD  KHV  MSL  PEL  PJL  PUR  \\\n",
       "0           1030533  2.0  1.0  1.0  1.0  0.0  2.0  0.0  2.0  1.0  0.0  2.0   \n",
       "1           1030773  2.0  1.0  1.0  1.0  0.0  2.0  0.0  2.0  1.0  0.0  2.0   \n",
       "55          4960359  0.0  1.0  1.0  0.0  1.0  5.0  0.0  2.0  3.0  0.0  2.0   \n",
       "56          4960603  0.0  1.0  1.0  0.0  1.0  5.0  0.0  2.0  3.0  0.0  2.0   \n",
       "57          4971420  0.0  1.0  1.0  0.0  1.0  5.0  0.0  2.0  3.0  0.0  2.0   \n",
       "\n",
       "Population  YRI  \n",
       "0           1.0  \n",
       "1           1.0  \n",
       "55          0.0  \n",
       "56          0.0  \n",
       "57          0.0  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asw_samples = node_pop_df[node_pop_df[\"ASW\"]>0] # filter for only nodes present in the ASW sample\n",
    "\n",
    "# there's no point in sorting in this specific case as there's only 1 ASW sample, but keeping for reference.\n",
    "asw_samples.sort_values(by=[\"ASW\"], ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "366ea5b3",
   "metadata": {},
   "source": [
    "Sadly, there's only 1 ASW haplotype in the current pangenome. To get more, we'll need to use `vg` to map alignments to these nodes.\n",
    "\n",
    "The first step for this is to get a list of all the samples for the population we're interested in:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ff84c0b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>FamilyID</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>FatherID</th>\n",
       "      <th>MotherID</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Population</th>\n",
       "      <th>Superpopulation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2821</th>\n",
       "      <td>2357</td>\n",
       "      <td>NA19625</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2855</th>\n",
       "      <td>2367</td>\n",
       "      <td>NA19700</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2856</th>\n",
       "      <td>2367</td>\n",
       "      <td>NA19701</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2857</th>\n",
       "      <td>2367</td>\n",
       "      <td>NA19702</td>\n",
       "      <td>NA19700</td>\n",
       "      <td>NA19701</td>\n",
       "      <td>1</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2858</th>\n",
       "      <td>2368</td>\n",
       "      <td>NA19703</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2987</th>\n",
       "      <td>2494</td>\n",
       "      <td>NA20357</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2988</th>\n",
       "      <td>2494</td>\n",
       "      <td>NA20358</td>\n",
       "      <td>NA20356</td>\n",
       "      <td>NA20357</td>\n",
       "      <td>1</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2989</th>\n",
       "      <td>2495</td>\n",
       "      <td>NA20359</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2990</th>\n",
       "      <td>2495a</td>\n",
       "      <td>NA20362</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2991</th>\n",
       "      <td>2505</td>\n",
       "      <td>NA20412</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>ASW</td>\n",
       "      <td>AFR</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>74 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     FamilyID SampleID FatherID MotherID  Sex Population Superpopulation\n",
       "2821     2357  NA19625        0        0    2        ASW             AFR\n",
       "2855     2367  NA19700        0        0    1        ASW             AFR\n",
       "2856     2367  NA19701        0        0    2        ASW             AFR\n",
       "2857     2367  NA19702  NA19700  NA19701    1        ASW             AFR\n",
       "2858     2368  NA19703        0        0    1        ASW             AFR\n",
       "...       ...      ...      ...      ...  ...        ...             ...\n",
       "2987     2494  NA20357        0        0    2        ASW             AFR\n",
       "2988     2494  NA20358  NA20356  NA20357    1        ASW             AFR\n",
       "2989     2495  NA20359        0        0    2        ASW             AFR\n",
       "2990    2495a  NA20362        0        0    1        ASW             AFR\n",
       "2991     2505  NA20412        0        0    2        ASW             AFR\n",
       "\n",
       "[74 rows x 7 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asw_samples = sample_data[sample_data[\"Population\"]==\"ASW\"]\n",
    "\n",
    "asw_samples"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
